### ISSUE TITLE
Idea: Graphics.ini For Organization And Graphics Deduplication

### ISSUE THREAD

**Issue by ChristianSilvermoon** [Wohlstand/TheXTech#15]
The files `music.ini` and `sounds.ini` allow customization of music and sounds respectively.

What if a Level or Episode directory (IE: `level0` corresponding to `level0.lvlx`) could contain a `graphics.ini` which defined the paths for particular graphic assets starting from the Episode Root Directory?

This could be used for:
- Better organization of the Episode Root Directory
- Asset Deduplication for when an Episode requires the same assets across multiple, but not all levels.

Example file structure:
```
Episode Root
├── graphics
│   └── custom_char_sprites_small_state.png
├── level0
│   └── graphics.ini
├── level0.lvlx
├── level1
│   └── graphics.ini
├── level1.lvlx
├── level3.lvlx
└── world.wldx
```

Example `graphics.ini` entry:
```
[mario-1]
name="Small Custom Player"
file="graphics/custom_char_sprites_small_state.png"
```

**Comment by Wohlstand** [comment]
I think, this idea I'll also apply to PGE Engine too! :+1:

**Comment by ChristianSilvermoon** [comment]
Friggin' sweet! ^^
That's awesome!

**Comment by 0lhi** [comment]
I think we should reconsider whether this idea is worth implementing.

The reason: While _music.ini_ and _sound.ini_ manage about 180 Soundfiles, _graphics.ini_ would have to manage **~2300 PNG Graphics**. Some of the Episodes I work on Repacking for TheXTech come with their own default Custom Graphics. When repacking those, I run a Command that deletes Graphics which are duplicated from the standard assets. Accounting for that in a _graphics.ini_ would be a severe pain and an unreasonable waste of time.

The request had two purposes:

- Asset De-Duplication
- Improve Organization

The De-Duplication can be done with 7z Archives once #132 is implemented.

For Improving the Organization, I'd like to suggest what SMBX2 is doing:

> Just for the record: Beta 4 features episode-wide graphics folders (**structured exactly like the base graphics folder**). Putting graphics in them makes SMBX2 **load them instead** of the basegame graphics, meaning that episode-wide custom graphics **don't unnecessarily contribute to load times**. - [Enjl](https://www.supermariobrosx.org/forums/viewtopic.php?p=370661#p370661)

Basically, if you download an episode like _Yoshi's Archipelago_ or _Talking Time Bros. 2.5_, you move it's graphics folder inside the 'Episode Name' Folder, and then the Episode Folder inside SMBX2's world folder. PNGs/GIFs that are not present in the Episodes' graphics folder will be loaded from the default assets.

Combining these two approaches could make _graphics.ini_ obsolete, would be easier to implement, easier to maintain, and it would follow an intuitive, established standard that Level Designers are already familiar with.

**Comment by ChristianSilvermoon** [comment]
> I think we should reconsider whether this idea is worth implementing.
> 
> The reason: While _music.ini_ and _sound.ini_ manage about 180 Soundfiles, _graphics.ini_ would have to manage **~2300 PNG Graphics**. Some of the Episodes I work on Repacking for TheXTech come with their own default Custom Graphics. When repacking those, I run a Command that deletes Graphics which are duplicated from the standard assets. Accounting for that in a _graphics.ini_ would be a severe pain and an unreasonable waste of time.
> 
> The request had two purposes:
> 
>     * Asset De-Duplication
> 
>     * Improve Organization
> 
> 
> The De-Duplication can be done with 7z Archives once #132 is implemented.
> 
> For Improving the Organization, I'd like to suggest what SMBX2 is doing:
> 
> > Just for the record: Beta 4 features episode-wide graphics folders (**structured exactly like the base graphics folder**). Putting graphics in them makes SMBX2 **load them instead** of the basegame graphics, meaning that episode-wide custom graphics **don't unnecessarily contribute to load times**. - [Enjl](https://www.supermariobrosx.org/forums/viewtopic.php?p=370661#p370661)
> 
> Basically, if you download an episode like _Yoshi's Archipelago_ or _Talking Time Bros. 2.5_, you move it's graphics folder inside the 'Episode Name' Folder, and then the Episode Folder inside SMBX2's world folder. PNGs/GIFs that are not present in the Episodes' graphics folder will be loaded from the default assets.
> 
> Combining these two approaches could make _graphics.ini_ obsolete, would be easier to implement, easier to maintain, and it would follow an intuitive, established standard that Level Designers are already familiar with.

I have to agree, actually :P

**Comment by Wohlstand** [comment]
Speaking about de-duplication: even now the game does the loading of only graphics that will be found while checking the stuff (level/world data directory/episode directory/default directory). So, no duplicates in the memory. Moondust Engine does the same since its foundation in 2014 with a little difference: TheXTech loads all files into memory but keeps them all compressed until the first request. Moondust Engine doesn't load these graphics at all, and after a first request, loads graphics from the disk, which makes slower time because of disk queries during the gameplay which affects low-end computers, but, keeps RAM usage lesser.

**Comment by 0lhi** [comment]
>Speaking about de-duplication: even now the game does the loading of only graphics that will be found while checking the stuff (level/world data directory/episode directory/default directory). So, no duplicates in the memory.

@Wohlstand That's cool! In that case, the SMBX2 structure should be super easy to implement, right? Just add _/episode/graphics/*_ to the stuff to be checked.

Just checked the priorities in SMBX2, they go like this:

1. Level Folder
2. Episode Folder Root
3. Episode Folder /graphics
4. Root /graphics

**Comment by ds-sloth** [comment]
@Wohlstand we've been talking some about the texture loading / decompression code recently, and that seems closely related to this issue.

In our codebase, across all platforms, we currently have two graphics load strategies:
(1) load all texture sizes and file paths to RAM on launch, load (and decompress if applicable) textures later
(2) load all textures and texture sizes to RAM on launch, decompress later

Strategy (1) is currently accelerated using a graphics load list file. This acts somewhat like an `ini` but unlike `IniProcessor`s it does not require everything to be allocated in memory at once. (Because of the hash-map / string structure of `IniProcessor`, that could be close to 4 MB -- quite a lot on our weakest platforms.) If one of us extends `IniProcessor` to load a single group at a time (preferably re-using a single heap-allocated hash-map) and use callbacks, then I think it would be great to replace this with an `ini`. In that case, we should discuss some about the ideal format and which fields need to be implemented.

Strategy (2) doesn't gain much performance by using a graphics load list, but as @ChristianSilvermoon noted it would add to flexibility.

@0lhi 's suggestion to add SMBX2 priorities seems reasonable to me also. It shouldn't impact performance on targets with slow filesystems because those would mostly use `graphics.ini` / `graphics.list` lists anyway.
